affinity/cpp-20/d0796r2.md

```diff
@@ -152,17 +152,59 @@ From a historic perspective, programming models for traditional high-performance
+
+  1. Holding a reference to the *execution resource* would prevent parallel execution from shutting down, thus (potentially) deadlocking the program.
+  2. Not all kinds of *resources* may have lifetimes that fit reference counting semantics. Some kinds of GPU *resources* only exist during parallel execution, for example; those *resources* cannot be valid if they escape the parallel region. In general, programming models that let a "host" processor launch code on a "different processor" have this issue.
+  3. Reference counting could have unattractive overhead if accessed concurrently, especially if code wants to traverse a particular subset of the *resource topology* inside a parallel region (e.g., to access GPU scratch memory).
+  4. Since users can construct arbitrary data structures from *resources* in a *resource hierarchy*, the proposal would need another *resource* type analogous to `weak_ptr`, in order to avoid circular dependencies that could prevent releasing *resources*.
+  5. There is no type currently in the Standard that has reference-counting semantics, but does not have `shared_` in its name (e.g., `shared_ptr` and `shared_future`). Adding a type like this sets a bad precedent for types with hidden costs and correctness issues (see (4)).
+
+#### What does validity of a resource mean?
+
+Here, we elaborate on what it means for a *resource* to be "valid." This proposal lets users encounter a *resource* either while traversing the *resource topology*, or through a *context* that uses the *resource*. "Viewing" the *resource* in the *resource topology* implies a lower level of "commitment" or "permanence" than using the *resource* in a *context*. In particular, 
+
+  1. It is implementation defined whether any subset of the *resource topology* reflects the current state of the *system*, or just a "snapshot." Ability to iterate a *resource*'s children in the *resource topology* need not imply ability to create a *context* from that *resource*. This may even vary between subsets of the *resource topology*. 
+  2. *Context* creation asserts *resource* validity.
+  3. Use of a *context* to launch parallel execution asserts *resource* validity.
+
+Here is a concrete example. Suppose that company "Aleph" makes an accelerator that can be viewed as a *resource*, and that has its own child *resources*. Users must call `Aleph_initialize()` in order to see the accelerator and its children as *resources* in the *resource topology*. Users must call `Aleph_finalize()` when they are done using the accelerator. 
+
+Questions:
+
+  1. What should happen if users are traversing the *resource topology*, but never use the accelerator's *resource* (other than to iterate past it), and something else concurrently calls `Aleph_finalize()`?
```

Comment:
You've raised some interesting questions here.

I think the best way to solve this would be to require that `this_system::get_resource` be thread-safe and initialise and finalize any third party or OS resources when constructing the topology before returning. This would mean that there's no way to concurrently interfere with how the implementation uses those resources without using them directly through their respective API, though that would have to be undefined behaviour. This also means that the topology would be a snapshot of the resources at that point, and don't actually require any of the underlying resources to be alive until they are actually used, i.e. when passed to the construction of a context.

The disadvantage of this approach is that the initial call to `this_system::get_resource` could be quite costly, depending on the implementation, though I'd expect that you would only do this once on startup or at specific intervals between computations.

Alternatively, if we go down the route that was suggested at Rapperswil of having a visitor approach to topology discovery, where the user provides some kind of visitor that traverses the topology and returns the resources that it wants to use. With this kind of approach, it might be worth having the `execution_resource` have active resources. Perhaps this pattern could construct the context directly?

It may also be worth having a thread-safe `valid` member function on the `execution_resource` to check if the underlying resource is still available. Though I think this would have to be named differently to represent that it is not simply a getter and that it must do some querying or initialising of resources within the topology to identify if the resource is still valid. However, even with this, you could still have the situation where a resource becomes unavailable between calling `valid` and constructing an execution context or after the context is constructed.

Perhaps instead we should say that you can always construct a context from an `execution_resource`, even if it's unavailable, but have the context handle the case where the resource is unavailable, by cancelling enqueued work and disallowing new submissions and throwing an exception or calling a callback function.

Reply:
> This also means that the topology would be a snapshot of the resources at that point, and don't actually require any of the underlying resources to be alive until they are actually used, i.e. when passed to the construction of a context.

I like this.  This could even solve the problem of some resources only being available inside a parallel region (like a GPU).  Topology is a snapshot of everything reachable from the root.

Reply:
> This would mean that there's no way to concurrently interfere with how the implementation uses those resources without using them directly through their respective API, though that would have to be undefined behaviour.

I'm OK with that being UB in general, possibly with specific APIs giving stronger guarantees.

Reply:
> though I'd expect that you would only do this once on startup or at specific intervals between computations.

That's right -- this is really about constructing a thread pool.  Crazy high-frequency dynamic hardware load balancing feels like it wants a different interface.  (Users might just see that as a special kind of resource -- it doesn't feel like something that naturally has an easily walkable topology.)

Reply:
> Alternatively, if we go down the route that was suggested at Rapperswil of having a visitor approach to topology discovery, where the user provides some kind of visitor that traverses the topology and returns the resources that it wants to use.

Was that in the notes?  I'd like to learn more about that.  (Sorry I missed that discussion at Rapperswil.)  I'm not sure, without knowing more at least, whether that would solve the problems we've discussed here.

Reply:
> Perhaps instead we should say that you can always construct a context from an execution_resource, even if it's unavailable, but have the context handle the case where the resource is unavailable, by cancelling enqueued work and disallowing new submissions and throwing an exception or calling a callback function.

Why not just attempt to create the context?  If the resource is no longer available, the context creation fails.  Otherwise, the context assumes the responsibility for keeping the resource alive or otherwise handling the case where the resource ceases to be alive at some point.

Reply:
@AerialMantis I reviewed the Rapperswil notes, and noticed that the affinity paper was marked "DND."  It looks like it wasn't covered there.  Were you referring to the P0443 discussion in SG1 (and/or joint with LEWG)?  [Here is the link](http://wiki.edg.com/bin/view/Wg21rapperswil2018/P0443) (requires usual login).  For example: "JB: My takeaway is that (P0443) executors are copyable things with reference semantics."

Reply:
Note: Discussion a week or two ago concluded that the affinity paper was actually discussed in SG1 in Rapperswil, but didn't make it into the minutes.

Reply:
(Note to self: I think I've addressed this in my local changes.)